File: 74-search-a-2d-matrix/search-a-2d-matrix.cpp
class Solution {
public:
    bool binarysearch(vector<int>arr,int target){
        int n=arr.size();
        int st=0;
        int end=n-1;
        while(st<=end){
            int mid=(st+end)/2;
            if(arr[mid]==target) return true;
            else if(arr[mid]>target) end=mid-1;
            else st=mid+1;
        }
        return false;
    }
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m=matrix.size();
        int n=matrix[0].size();
        int st=0;
        int end=m-1;
        int mid=(st+end)/2;
        int row=-1;
        while(st<=end){
             mid=(st+end)/2;
             if (matrix[mid][0] <= target && target <= matrix[mid][n - 1]) {
                row = mid;
                break;
            }
            else if(matrix[mid][0]>target){
            //    bool check= binarysearch(matrix[mid],target);
            //    if(check==true) return true;
               end=mid-1;   
            } 
            else st=mid+1;
        }
        if(row==-1)return false;
       return binarysearch(matrix[mid],target);
       // return false;
    }
};

File: 74-search-a-2d-matrix/search-a-2d-matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        
        int n=matrix.length-1;
        int m=matrix[0].length-1;
        if(n==0 && matrix[0][0]==target)return true;
        int i=0,j=m;
        while(i<=n && j>=0){
            if(matrix[i][j]==target) return true;
            else if(matrix[i][j]>target) j--;
            else{
               i++;
            }
            
        }
        return false;

    }
}